File: grin/R/criterion.R
# criterion.R: stopping decisions for adaptive designs.
#
# Port of grintools' Criterion/Target/Decision (Python). The EXPERIMENTER declares
# a Criterion from Targets; GRIN evaluates it. Three ways to use it:
#
#   1. precision    grin_target_precision(params=..., sd_max=... | ci_width_max=...)
#                   stop when the parameter posterior is tight enough. This is the
#                   "I want the space measured, verdict aside" target.
#   2. probability  grin_target_probability("PI"|"PS_A"|"PS_B"|"*_violated", at_least=0.9)
#                   stop when a construct probability crosses a threshold. This is
#                   the "I want the verdict" target.
#   3. combine      grin_criterion(list(...targets...), combine = "all"|"any")
#
# Probability targets read the construct list's evidence_* flags. When an evidence
# flag is FALSE, the target is reported as unreachable (Decision$blocked_by), which
# is how the PI identifiability limit surfaces: a threshold on a construct the data
# cannot decide will never be met, and we say so rather than looping forever.

.CONSTRUCT_MAP <- list(
  PI            = list(key = "p_PI",    evkey = "evidence_PI",    violated = FALSE),
  PS_A          = list(key = "p_sep_A", evkey = "evidence_sep_A", violated = FALSE),
  PS_B          = list(key = "p_sep_B", evkey = "evidence_sep_B", violated = FALSE),
  PI_violated   = list(key = "p_PI",    evkey = "evidence_PI",    violated = TRUE),
  PS_A_violated = list(key = "p_sep_A", evkey = "evidence_sep_A", violated = TRUE),
  PS_B_violated = list(key = "p_sep_B", evkey = "evidence_sep_B", violated = TRUE)
)

#' @keywords internal
.select_indices <- function(params) {
  if (is.null(params) || identical(params, "all")) return(seq_along(PARAM_NAMES))
  idx <- integer(0)
  for (p in params) {
    if (p %in% names(PARAM_GROUPS)) {
      idx <- c(idx, PARAM_GROUPS[[p]])
    } else if (p %in% PARAM_NAMES) {
      idx <- c(idx, match(p, PARAM_NAMES))
    } else {
      stop(sprintf("unknown parameter selector '%s'; use a name in {%s} or a group in {%s}",
                   p, paste(PARAM_NAMES, collapse = ", "), paste(names(PARAM_GROUPS), collapse = ", ")),
          call. = FALSE)
    }
  }
  sort(unique(idx))
}

#' Precision stopping target
#'
#' Stop when the posterior for the selected parameters is tight enough. Exactly one
#' of `sd_max`/`ci_width_max` must be given.
#'
#' @param params Character vector of parameter names ([PARAM_NAMES]) and/or groups
#'   (`"zx"`, `"zy"`, `"rho"`), or `NULL`/`"all"` for every parameter.
#' @param sd_max Stop when every selected parameter's posterior SD is at most this.
#' @param ci_width_max Stop when every selected parameter's 90% CI width is at most this.
#' @export
grin_target_precision <- function(params = NULL, sd_max = NULL, ci_width_max = NULL) {
  if (is.null(sd_max) == is.null(ci_width_max)) {
    stop("precision target needs exactly one of sd_max or ci_width_max", call. = FALSE)
  }
  structure(list(kind = "precision",
                cfg = list(params = params, sd_max = sd_max, ci_width_max = ci_width_max)),
           class = "grin_target")
}

#' Construct-probability stopping target
#'
#' Stop when a construct probability crosses a threshold.
#'
#' @param construct One of `"PI"`, `"PS_A"`, `"PS_B"`, or a `"*_violated"` complement.
#' @param at_least Probability threshold.
#' @export
grin_target_probability <- function(construct, at_least) {
  if (!(construct %in% names(.CONSTRUCT_MAP))) {
    stop(sprintf("unknown construct '%s'; choose from {%s}", construct,
                paste(names(.CONSTRUCT_MAP), collapse = ", ")), call. = FALSE)
  }
  structure(list(kind = "probability",
                cfg = list(construct = construct, at_least = as.numeric(at_least))),
           class = "grin_target")
}

#' @keywords internal
.target_check <- function(target, result, constructs) {
  cfg <- target$cfg
  if (identical(target$kind, "precision")) {
    idx <- .select_indices(cfg$params)
    names_ <- if (!is.null(result$names)) result$names else PARAM_NAMES
    if (!is.null(cfg$sd_max)) {
      vals <- as.numeric(result$std)[idx]; thr <- cfg$sd_max; q <- "sd"
    } else {
      vals <- (as.numeric(result$ci_high) - as.numeric(result$ci_low))[idx]
      thr <- cfg$ci_width_max; q <- "ci_width"
    }
    worst <- which.max(vals)
    return(list(met = all(vals <= thr), value = vals[worst],
               name = sprintf("%s:%s", q, names_[idx[worst]]), threshold = thr,
               reachable = TRUE, note = ""))
  }
  # probability target
  cname <- cfg$construct; thr <- cfg$at_least
  if (is.null(constructs)) {
    return(list(met = FALSE, value = NaN, name = cname, threshold = thr,
               reachable = NA, note = "no constructs supplied; cannot evaluate"))
  }
  spec <- .CONSTRUCT_MAP[[cname]]
  p <- as.numeric(constructs[[spec$key]])
  if (spec$violated) p <- 1.0 - p
  evkey_val <- constructs[[spec$evkey]]
  ev <- if (is.null(evkey_val)) TRUE else isTRUE(evkey_val)
  note <- if (ev) "" else paste("evidence flag is False: the data may not be able to decide",
                                "this construct in the current regime (a property of the",
                                "data, not of GRIN)")
  list(met = (p >= thr) && ev, value = p, name = cname, threshold = thr,
      reachable = ev, note = note)
}

#' Combine stopping targets into a criterion
#'
#' @param targets A list of `grin_target` objects (see [grin_target_precision()],
#'   [grin_target_probability()]).
#' @param combine `"all"` (default, stop only once every target is met) or `"any"`.
#' @export
grin_criterion <- function(targets, combine = "all") {
  if (!(combine %in% c("all", "any"))) stop("combine must be 'all' or 'any'", call. = FALSE)
  structure(list(targets = targets, combine = combine), class = "grin_criterion")
}

#' @keywords internal
.criterion_needs_constructs <- function(criterion) {
  any(vapply(criterion$targets, function(t) identical(t$kind, "probability"), logical(1)))
}

#' Evaluate a stopping criterion against an inference result
#'
#' @param criterion A `grin_criterion` (see [grin_criterion()]).
#' @param result A `grin_result` (e.g. `grin_infer(...)$result`).
#' @param constructs Optional constructs list (e.g. `grin_infer(...)$constructs`);
#'   required if any target in `criterion` is a probability target.
#' @return A `grin_decision`: `$stop` (logical), `$checks` (list of per-target detail),
#'   `$combine`, `$blocked_by` (names of targets whose evidence flag says the data
#'   cannot decide them -- these will never be met, no matter how much data you add).
#' @examples
#' \donttest{
#' M <- matrix(c(71, 17,  9,  5,
#'               20, 67,  5,  9,
#'               13,  6, 63, 20,
#'                5, 10, 15, 71), nrow = 4, byrow = TRUE)
#' out <- grin_infer(M)
#' crit <- grin_criterion(list(
#'   grin_target_precision(params = c("zx", "zy"), sd_max = 0.10),
#'   grin_target_probability("PS_A", at_least = 0.90)
#' ), combine = "any")
#' decision <- grin_evaluate(crit, out$result, out$constructs)
#' print(decision)
#' }
#' @export
grin_evaluate <- function(criterion, result, constructs = NULL) {
  checks <- lapply(criterion$targets, .target_check, result = result, constructs = constructs)
  mets <- vapply(checks, function(c) isTRUE(c$met), logical(1))
  stop_ <- if (identical(criterion$combine, "all")) all(mets) else any(mets)
  blocked <- vapply(checks, function(c) identical(c$reachable, FALSE), logical(1))
  blocked_names <- vapply(checks[blocked], function(c) c$name, character(1))
  structure(list(stop = stop_, checks = checks, combine = criterion$combine,
                blocked_by = blocked_names),
           class = "grin_decision")
}

#' @export
print.grin_decision <- function(x, ...) {
  cat(sprintf("stop = %s  (combine = '%s')\n", x$stop, x$combine))
  for (c in x$checks) {
    mark <- if (isTRUE(c$met)) "met" else "not met"
    val <- if (is.nan(c$value)) "nan" else sprintf("%.3f", c$value)
    cat(sprintf("    %-16s %s vs %.3f  [%s]\n", c$name, val, c$threshold, mark))
    if (nzchar(c$note)) cat(sprintf("        note: %s\n", c$note))
  }
  if (length(x$blocked_by) > 0) {
    cat(sprintf("    unreachable target(s): %s (threshold may never be met with current data)\n",
               paste(x$blocked_by, collapse = ", ")))
  }
  invisible(x)
}

#' Convenience: evaluate a single precision target
#' @inheritParams grin_target_precision
#' @param result A `grin_result` (e.g. `grin_infer(...)$result`).
#' @export
grin_stop_on_precision <- function(result, sd_max = NULL, ci_width_max = NULL, params = NULL) {
  grin_evaluate(grin_criterion(list(grin_target_precision(params = params, sd_max = sd_max,
                                                          ci_width_max = ci_width_max))),
               result)
}

File: packages/grintools/grintools/criterion.py
"""
criterion.py: stopping decisions for adaptive designs.

Split out of io.py so `grintools.criterion` can be imported (and reasoned about)
independently of the input-normalisation layer; `Target`/`Criterion` only need
PARAM_NAMES/PARAM_GROUPS from io.py, not any of its parsing machinery.
"""
from __future__ import annotations
import numpy as np

from .io import PARAM_NAMES, PARAM_GROUPS

# =========================================================================== #
# Stopping decisions. The EXPERIMENTER declares a Criterion from Targets; GRIN
# evaluates it. Four ways to use it:
#
#   1. passthrough  : build no Criterion; read result.* and constructs yourself.
#   2. precision    : Target.precision(params=..., sd_max=... | ci_width_max=...)
#                     stop when the parameter posterior is tight enough. This is
#                     the "I want the space measured, verdict aside" target.
#   3. probability  : Target.probability("PI"|"PS_A"|"PS_B"|"*_violated", at_least=0.9)
#                     stop when a construct probability (from model_posterior)
#                     crosses a threshold. This is the "I want the verdict" target.
#   4. combine      : Criterion([...targets...], combine="all"|"any")
#
# The construct targets read model_posterior's per-matrix dict directly, including
# its evidence_* flags. When an evidence flag is False the target is reported as
# unreachable (Decision.blocked_by), which is how the PI identifiability limit
# surfaces: a threshold on a construct the data cannot decide will never be met,
# and we say so rather than letting the loop run forever.
# =========================================================================== #
_CONSTRUCT_MAP = {
    "PI":            ("p_PI",    "evidence_PI",    False),
    "PS_A":          ("p_sep_A", "evidence_sep_A", False),
    "PS_B":          ("p_sep_B", "evidence_sep_B", False),
    "PI_violated":   ("p_PI",    "evidence_PI",    True),
    "PS_A_violated": ("p_sep_A", "evidence_sep_A", True),
    "PS_B_violated": ("p_sep_B", "evidence_sep_B", True),
}


def _select_indices(params):
    if params is None or params == "all":
        return list(range(len(PARAM_NAMES)))
    idx = []
    for p in ([params] if isinstance(params, str) else list(params)):
        if p in PARAM_GROUPS:
            idx.extend(PARAM_GROUPS[p])
        elif p in PARAM_NAMES:
            idx.append(PARAM_NAMES.index(p))
        else:
            raise ValueError(f"unknown parameter selector {p!r}; use a name in "
                             f"{PARAM_NAMES} or a group in {list(PARAM_GROUPS)}")
    return sorted(set(idx))


class Target:
    def __init__(self, kind, cfg):
        self.kind = kind
        self.cfg = cfg

    @classmethod
    def precision(cls, *, params=None, sd_max=None, ci_width_max=None):
        if (sd_max is None) == (ci_width_max is None):
            raise ValueError("precision target needs exactly one of sd_max or ci_width_max")
        return cls("precision", dict(params=params, sd_max=sd_max, ci_width_max=ci_width_max))

    @classmethod
    def probability(cls, construct, *, at_least):
        if construct not in _CONSTRUCT_MAP:
            raise ValueError(f"unknown construct {construct!r}; choose from {list(_CONSTRUCT_MAP)}")
        return cls("probability", dict(construct=construct, at_least=float(at_least)))

    def check(self, result, constructs):
        if self.kind == "precision":
            idx = _select_indices(self.cfg["params"])
            names = getattr(result, "names", PARAM_NAMES)
            if self.cfg["sd_max"] is not None:
                vals = np.asarray(result.std, float)[idx]; thr = self.cfg["sd_max"]; q = "sd"
            else:
                vals = (np.asarray(result.ci_high, float) - np.asarray(result.ci_low, float))[idx]
                thr = self.cfg["ci_width_max"]; q = "ci_width"
            worst = int(np.argmax(vals))
            return dict(met=bool(np.all(vals <= thr)), value=float(vals[worst]),
                        name=f"{q}:{names[idx[worst]]}", threshold=thr, reachable=True, note="")
        # probability
        c = self.cfg["construct"]; thr = self.cfg["at_least"]
        if constructs is None:
            return dict(met=False, value=float("nan"), name=c, threshold=thr,
                        reachable=None, note="no constructs supplied; cannot evaluate")
        key, evkey, violated = _CONSTRUCT_MAP[c]
        p = float(constructs[key])
        if violated:
            p = 1.0 - p
        ev = bool(constructs.get(evkey, True))
        note = "" if ev else ("evidence flag is False: the data may not be able to decide "
                              "this construct in the current regime (a property of the data, "
                              "not of GRIN)")
        return dict(met=bool(p >= thr and ev), value=p, name=c, threshold=thr,
                    reachable=ev, note=note)


class Decision:
    def __init__(self, stop, checks, combine, blocked):
        self.stop = bool(stop)
        self.checks = checks
        self.combine = combine
        self.blocked_by = blocked

    def __bool__(self):
        return self.stop

    def summary(self):
        lines = [f"stop = {self.stop}  (combine='{self.combine}')"]
        for c in self.checks:
            mark = "met" if c["met"] else "not met"
            val = "nan" if c["value"] != c["value"] else f"{c['value']:.3f}"
            lines.append(f"    {c['name']:16s} {val} vs {c['threshold']:.3f}  [{mark}]")
            if c["note"]:
                lines.append(f"        note: {c['note']}")
        if self.blocked_by:
            lines.append(f"    unreachable target(s): {self.blocked_by} "
                         f"(threshold may never be met with current data)")
        return "\n".join(lines)


class Criterion:
    def __init__(self, targets, combine="all"):
        self.targets = list(targets)
        if combine not in ("all", "any"):
            raise ValueError("combine must be 'all' or 'any'")
        self.combine = combine

    @property
    def needs_constructs(self):
        return any(t.kind == "probability" for t in self.targets)

    def evaluate(self, result=None, constructs=None):
        checks = [t.check(result, constructs) for t in self.targets]
        mets = [c["met"] for c in checks]
        stop = all(mets) if self.combine == "all" else any(mets)
        blocked = [c["name"] for c in checks if c["reachable"] is False]
        return Decision(stop, checks, self.combine, blocked)


def stop_on_precision(result, *, sd_max=None, ci_width_max=None, params=None):
    """Convenience for the common single-target precision case."""
    return Criterion([Target.precision(params=params, sd_max=sd_max,
                                       ci_width_max=ci_width_max)]).evaluate(result)
